iwyu_ast_util.cc

```diff
@@ -846,9 +846,14 @@ const NamedDecl* GetDefinitionAsWritten(const NamedDecl* decl) {
     if (const ClassTemplateDecl* tpl_decl = DynCastFrom(decl))
       decl = tpl_decl->getTemplatedDecl();  // convert back to CXXRecordDecl
   } else if (const FunctionDecl* func_decl = DynCastFrom(decl)) {
-    if (const FunctionDecl* tpl_pattern =
-        func_decl->getTemplateInstantiationPattern())
-      decl = tpl_pattern;
+    // If we're instantiated from a template, use the template pattern as the
+    // decl-as-written.
+    // But avoid friend declarations in templates, something happened in Clang
+    // r283207 that caused them to form a dedicated redecl chain, separate
+    // from all other redecls.
+    const FunctionDecl* tp_decl = func_decl->getTemplateInstantiationPattern();
```

Comment:
The only concern is `tp_decl` naming. Should it be `tpl` instead? But we already have `const ClassTemplateDecl* tpl_decl` above so maybe keeping `tpl_pattern` is better but it can be causing more line wrapping.